core/src/main/java/org/infinispan/CacheStream.java

```diff
@@ -0,0 +1,251 @@
+    * <p>This method is ran distributed by default with a distributed backing cache.  However if you wish for this
+    * operation to run locally you can use the {@link CacheStream#iterator()} method to return all of the results
+    * locally and then use {@link Iterator#forEachRemaining(Consumer)} method for a single threaded variant.  If you
+    * wish to have a parallel variant you can use {@link java.util.stream.StreamSupport#stream(Spliterator, boolean)}
+    * passing in the spliterator from the stream.  In either case remember you <b>must</b> close the stream after
+    * you are done processing the iterator or spliterator..</p>
+    * @param action
+    */
+   @Override
+   void forEach(Consumer<? super R> action);
+
+   /**
+    * {@inheritDoc}
+    * <p>Usage of this operator requires closing this stream after you are done with the iterator.  The preferred
+    * usage is to use a try with resource block on the stream.</p>
+    * <p>This method has special usage with the {@link org.infinispan.CacheStream.SegmentCompletionListener} in
+    * that as entries are retrieved from the next method it will complete segments.</p>
+    * <p>This method obeys the {@link CacheStream#distributedBatchSize(int)} setting by only ever returning the
+    * elements that mapped to that many keys.  Note that when using methods such as
+    * {@link CacheStream#flatMap(Function)} that you will have possibly more than 1 element mapped to a given key
+    * so this doesn't guarantee that many number of entries are returned per batch.</p>
+    * @return the element iterator for this stream
+    */
+   @Override
+   Iterator<R> iterator();
+
+   /**
+    * {@inheritDoc}
+    * <p>Usage of this operator requires closing this stream after you are done with the spliterator.  The preferred
+    * usage is to use a try with resource block on the stream.</p>
+    * @return the element spliterator for this stream
+    */
+   @Override
+   Spliterator<R> spliterator();
+
+   /**
+    * {@inheritDoc}
+    * <p>This method has special usage when used with a distributed cache backing this set.  This operation will act
+    * as an intermediate iterator operation requiring data be brought locally for proper behavior.  This is
+    * described in more detail in the {@link CacheStream} documentation</p>
+    * <p>This intermediate iterator operation will be performed locally only requiring all elements to be in memory</p>
+    * @return the new stream
+    */
+   @Override
+   Stream<R> sorted();
+
+   /**
+    * {@inheritDoc}
+    * <p>This method has special usage when used with a distributed cache backing this set.  This operation will act
+    * as an intermediate iterator operation requiring data be brought locally for proper behavior.  This is
+    * described in more detail in the {@link CacheStream} documentation</p>
+    * <p>This intermediate iterator operation will be performed locally only requiring all elements to be in memory</p>
```

Comment:
I'm not sure about the "special usage" wording. I'd just say that this operation is never executed on remote nodes, and all the intermediate results have to be fetched on the originator before sorting can begin. Maybe also recommend that users limit sorting to small result sets.